core/src/main/java/org/elasticsearch/script/Script.java

```diff
@@ -19,281 +19,590 @@
-        boolean hasContentType = contentType != null;
-        out.writeBoolean(hasContentType);
-        if (hasContentType) {
-            XContentType.writeTo(contentType, out);
+
+        /**
+         * Validates the parameters and creates an {@link Script}.
+         * @param lang The default lang is not a compile-time constant and must be provided
+         *             at run-time this way in case a legacy default language is used from
+         *             previously stored queries.
+         */
+        private Script build(String lang) {
+            if (type == null) {
+                throw new IllegalArgumentException(
+                    "must specify either code for an [" + ScriptType.INLINE.getParseField().getPreferredName() + "] script " +
+                        "or an id for a [" + ScriptType.STORED.getParseField().getPreferredName() + "] script " +
+                        "or [" + ScriptType.FILE.getParseField().getPreferredName() + "] script");
+            }
+
+            if (idOrCode == null) {
+                throw new IllegalArgumentException("must specify an id or code for a script");
+            }
+
+            if (options.size() > 1 || options.size() == 1 && options.get(CONTENT_TYPE_OPTION) == null) {
+                throw new IllegalArgumentException("illegal compiler options [" + options + "] specified");
+            }
+
+            return new Script(type, this.lang == null ? lang : this.lang, idOrCode, options, params);
         }
     }
 
+    private static final ObjectParser<Builder, ParseFieldMatcherSupplier> PARSER = new ObjectParser<>("script", Builder::new);
+
+    static {
+        // Defines the fields necessary to parse a Script as XContent using a ConstructingObjectParser.
+        PARSER.declareField(Builder::setInline, parser -> parser, ScriptType.INLINE.getParseField(), ValueType.OBJECT_OR_STRING);
+        PARSER.declareString(Builder::setStored, ScriptType.STORED.getParseField());
+        PARSER.declareString(Builder::setFile, ScriptType.FILE.getParseField());
+        PARSER.declareString(Builder::setLang, LANG_PARSE_FIELD);
+        PARSER.declareField(Builder::setOptions, XContentParser::mapStrings, OPTIONS_PARSE_FIELD, ValueType.OBJECT);
+        PARSER.declareField(Builder::setParams, XContentParser::map, PARAMS_PARSE_FIELD, ValueType.OBJECT);
+    }
+
     /**
-     * Method for getting the script.
-     * @return The cache key of the script to be compiled/executed.  For dynamic scripts this is the actual
-     *         script source code.  For indexed scripts this is the id used in the request.  For on disk scripts
-     *         this is the file name.
+     * Convenience method to call {@link Script#parse(XContentParser, ParseFieldMatcher, String)}
+     * using the default scripting language.
      */
-    public String getScript() {
-        return script;
+    public static Script parse(XContentParser parser, ParseFieldMatcher matcher) throws IOException {
```

Comment:
I'm inclined to leave this one simply because it's kind of the standard.  The one I would like to remove is the public static Script parse(XContentParser parser, ParseFieldMatcher matcher, String defaultLang)... that whole mess can go away once we don't have to support Groovy as a legacy default language.  Hopefully, 6.0.